### ISSUE TITLE
Failure message not in line with documentation

### ISSUE THREAD

**Issue by thomasthiry** [shouldly/shouldly#524]
I've been using Shouldly for some time now, but lately with my (dotnet core) projects and Shouldly 3.0.1 the failure messages are not as good as before. The name of the variable is not put anymore in the message.
for example for

```
employee.FirstName.ShouldBe("John"); // fails
--> message: "Dale" should be "John" but was not
```

When I expected something like this
```
employee.FirstName.ShouldBe("John"); // fails
--> message: FirstName should be "John" but was "Dale"
```

My question: is that on purpose or did I miss something? This is not in line with the documentation I think.
(it kinda diminishes my point when I try to convince people to use it :-) )

**Comment by thomasthiry** [comment]
It can be fixed by changing the debugging information of the test project to full:
Project properties > Build > Advanced... > Debugging Information = Full
Thanks @JosephWoodward for your support!

**Comment by JosephWoodward** [comment]
Glad you got to the bottom of it, thanks for updating the issue!

**Comment by thomasthiry** [comment]
Even though the debugging information seems to help, it doesn't solve the problem all the time.
I have a dotnet core project with this debugging information:
![image](https://user-images.githubusercontent.com/2093956/46574826-f1e87700-c9a9-11e8-9256-6ac251e118dc.png)
And yet, this xunit test

[Fact]
        public void Test1()
        {
            var ab = "aa";

            ab.ShouldBe("errr");
        }

gives this error message:

Shouldly.ShouldAssertException : Shouldly uses your source code to generate its great error messages, build your test project with full debug information to get better error messages
The provided expression
    should be
"errr"
    but was
"aa"
    difference
Difference     |  |    |    |    |   
               | \|/  \|/  \|/  \|/  
Index          | 0    1    2    3    
Expected Value | e    r    r    r    
Actual Value   | a    a              
Expected Code  | 101  114  114  114  
Actual Code    | 97   97

**Comment by thomasthiry** [comment]
Hi @JosephWoodward, do you think you'll find a solution to that or is it impossible? That would be so sad. :-)

**Comment by slang25** [comment]
For me the behaviour doesn't change between full and portable PDBs, but it will only work when I have a debugger attached.

I'm running a test via Rider and xUnit, it will fail if I run it, but if I step through it, I get the right message.

The stackframe walking stuff is about getting the file path and line number, so that we can read the assertion line from the source. We could use [`CallerLineNumberAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callerlinenumberattribute?view=netframework-4.7.2) and [`CallerFilePathAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callerfilepathattribute?view=netframework-4.7.2) to achieve the same result and pass them through to [here](https://github.com/shouldly/shouldly/blob/ea5739a3859627d4112723c1ddab557d1b842e1b/src/Shouldly/Internals/SourceCodeTextGetter.cs#L73).

**Comment by slang25** [comment]
After some experimenting, I'm finding that the above issue is only a problem when calling `new StackFrame(0)`, if you instead use `new StackTrace(true)` and walk the stack in that way, it works reliably. Expect a PR later this week 😃

**Comment by thomasthiry** [comment]
Great, looking forward to testing that! ;-)

**Comment by JosephWoodward** [comment]
@stuart-lang What does your project look like, it is anywhere on GitHub? I'm having a hard time replicating this now. 😕

**Comment by JosephWoodward** [comment]
Having looked at this further it seems that running Shouldly in a .NET Core console app will output the correct errors, but in an xunit project it doesn't. If I reference the Shouldly source directly in the xunit test then the problem goes away.

**Comment by slang25** [comment]
Using Shouldy 3.0.1 from nuget, for me it still fails for a .NET Core 2.1 console app, unless I debug it. Same behavior for xunit.
If I change my app or tests to Release, it still works (but sometimes is quirky due to inlining and stuff I guess).
It only works when I am debugging in Rider, or I have a console app / test project in the same solution as Shouldly.

**Comment by JosephWoodward** [comment]
@slang25 Oh really? Let me double check mine. So if I set it to debug in debug mode then it works as expected.

Further findings:

- Referencing a netstandard 1.3 Shouldly DLL in Debug mode doesn't work
- Referencing a netstandard 2.0 Shouldly DLL in Debug mode works

**Comment by slang25** [comment]
I'm getting a little confused, there Debug (vs Release) build configuration for Shouldly, then the same for the console app, then there's running it with the debugger attached.

For me under some scenarios it only works when the debugger is attached.

**Comment by JosephWoodward** [comment]
Sorry, just reread my original message and realised the first part made no sense. That's right, a console app when set to Debug build configuration and running with the debugger attached works as expected. 

It seems like in all cases (most?) attaching a debugger seems to work as expected,

**Comment by JosephWoodward** [comment]
Think I might have found the culprit, looks like it's [this section of code](https://github.com/shouldly/shouldly/blob/master/src/Shouldly/Internals/SourceCodeTextGetter.cs#L85) getting the wrong code lines and not being able to match the shouldly method due to release vs debug differences.

<img width="642" alt="screenshot 2018-10-27 at 08 38 52" src="https://user-images.githubusercontent.com/1237341/47601164-2c02d280-d9c4-11e8-99c9-c4b28bef5740.png">

Currently looking at ways of fixing it, was looking at the CallerLineNumberAttribute and CallerFilePathAttribute but being method attributes means they'd need to be added to every shouldly method I think so not sure how viable they'd be.

**Comment by thomasthiry** [comment]
Well done! Can't wait to have the fix. ;-)

**Comment by JosephWoodward** [comment]
Linking this with this issue raised: https://github.com/shouldly/shouldly/issues/412

**Comment by JosephWoodward** [comment]
It appears that FluentAssertions also has the same issue, however they output their assertions differently so it's less noticeable:

Example:

> Expected string to be "Boo" with a length of 3, but "Hello" has a length of 5.
vs 
> Expected string myVar to be "Boo" with a length of 3, but "Hello" has a length of 5.

**Comment by JosephWoodward** [comment]
Found it! Inlining on this method is causing the issue. No idea how though, soo many questions!

<img width="701" alt="screenshot 2018-11-01 at 01 18 08" src="https://user-images.githubusercontent.com/1237341/47827171-1dc20700-dd74-11e8-953b-c9969dca788c.png">

**Comment by JosephWoodward** [comment]
Okay, so it seems like you can remove the no inlining attribute if you change the code from this:

```csharp
        public static void ShouldBe(
            this string actual,
            string expected,
            Func<string> customMessage,
            StringCompareShould options)
        {
            var assertion = StringShouldBeAssertionFactory.Create(expected, actual, options);
            ExecuteAssertion(assertion, customMessage);
        }

        private static void ExecuteAssertion(Internals.Assertions.IAssertion assertion, Func<string> customMessage)
        {
            try
            {
                if (assertion.IsSatisfied()) return;
            }
            catch (ArgumentException ex)
            {
                throw new ShouldAssertException(ex.Message, ex);
            }
            throw new ShouldAssertException(assertion.GenerateMessage(customMessage()));
        }
```

to this:

```csharp
        public static void ShouldBe(
            this string actual,
            string expected,
            Func<string> customMessage,
            StringCompareShould options)
        {
            var assertion = StringShouldBeAssertionFactory.Create(expected, actual, options);

            try
            {
                if (assertion.IsSatisfied()) return;
            }
            catch (ArgumentException ex)
            {
                throw new ShouldAssertException(ex.Message, ex);
            }
            throw new ShouldAssertException(assertion.GenerateMessage(customMessage()));
        }
```
Basically, move the `ExecuteAssertion` method into the caller. 😕

**Comment by thomasthiry** [comment]
Nice, well done! :-) So I suppose we can expect a new version with a fix soon? I'll be happy to confirm it fixes the problem.

**Comment by JosephWoodward** [comment]
@thomasthiry Yup! Fix should be very soon, just try to find out if any other areas are affected.

**Comment by JosephWoodward** [comment]
Looking at the decompiled source, this is what it's doing, going to run some tests to see what part is causing it.

```csharp
	public static void ShouldBe(this string actual, string expected, Func<string> customMessage, StringCompareShould options)
	{
		ExecuteAssertion(StringShouldBeAssertionFactory.Create(expected, actual, options, "ShouldBe"), customMessage);
	}

	private static void ExecuteAssertion(IAssertion assertion, Func<string> customMessage)
	{
		try
		{
			if (assertion.IsSatisfied())
			{
				return;
			}
		}
		catch (ArgumentException ex)
		{
			throw new ShouldAssertException(ex.Message, ex);
		}
		throw new ShouldAssertException(assertion.GenerateMessage(customMessage()));
	}
```

**Comment by jnm2** [comment]
> Found it! Inlining on this method is causing the issue. No idea why though, soo many questions!

There are two things at play that are often mistaken for each other. One is inlining, the other is tail calling. Tail calls happen when the method returns immediately after calling another method. Either there's no return value or the return value is being immediately returned to the caller; either way, there's no need for the stack frame to stay around.

If optimizations are disabled, tail call optimization doesn't happen and the frame stays there even though it contains no useful information from the perspective of a stack frame. (The stack frame is all about pointing to where execution will continue next, not necessarily where it came from, to the extent that those two things diverge.)

**Comment by JosephWoodward** [comment]
@jnm2 I see, so in this instance it looks like it could be the tail calling that's interfering with the stack trace?

**Comment by slang25** [comment]
That would add up. What I saw is that when optimization were enabled, shouldly would not get the right results from walking the call stack. I would expect if anything adding the no inlining attribute to fix it, not the other way around.

**Comment by jnm2** [comment]
The only way I'm aware of to guarantee that a stack frame stays in place:

- make sure inlining is prevented
- make sure that some kind of work is being done after the last method call that cares what the stack looks like. (If you just do something useless after the last call though, there's a risk it'll be optimized away at runtime and tail calling will still happen.)

**Comment by JosephWoodward** [comment]
Thanks for your input on this @jnm2.

Moving the aforementioned method inside the caller fixed it in this instance, but there's nothing to stop this from coming back around.

One option I was considering is to release Shouldy unoptimised, seeing as it's a unit testing library optimisations isn't essential. What are your thoughts?

**Comment by jnm2** [comment]
Disabling optimizations is a good idea; I forgot, I have seen that done before.
Is there an automated way you could set up a test that would discover this issue in any Shouldly method the instant it was introduced?

**Comment by slang25** [comment]
I thought that optimizations were enabled/disabled, it resulted in an assembly attribute, however it's only the entrypoint assembly that takes effect at runtime, so basically we couldn't use that here as it depends on the consuming test project.

Appologies @JosephWoodward, I read that you were removing the no inlining attribute, not adding it (which makes so much more sense). That also fixed it for me, I had to add it to all of the overloads. I agreed there should be a more robust way of solving this.

**Comment by slang25** [comment]
Nope, talking bs, jit will respect the setting for each assembly 😃

**Comment by JosephWoodward** [comment]
> Nope, talking bs, jit will respect the setting for each assembly

What do you mean, I'm lost 😕

**Comment by JosephWoodward** [comment]
@stuart-lang By that PR I presume you approve of disabling optimisations? 😁

**Comment by slang25** [comment]
Sorry, what I meant was that I thought that when you run some dotnet code that the entrypoint project was what dictated weather JIT optimizations were off/on for the whole AppDomain. That isn't the case, each loaded assembly get's it's optimization flag respected.

**Comment by slang25** [comment]
I approve in the sense that it fixes this problem 😆 A better long term fix would be to have a smarter implementation.

**Comment by JosephWoodward** [comment]
> Sorry, what I meant was that I thought that when you run some dotnet code that the entrypoint project was what dictated weather JIT optimizations were off/on for the whole AppDomain. That isn't the case, each loaded assembly get's it's optimization flag respected.

Ah, I see what you mean. Yes, that's right - I verified it the other night, so disabling optimisations would solve it forever (though not the most elegant solution admittedly). Not sure what the difference in package size would be, or the hit on runtime performance.

**Comment by slang25** [comment]
Package size, zero impact. Basically it's an assembly attribute flag that is flipped, it's not a compiler thing, it's a JIT thing.

There would be a perf hit, but it doesn't seem likely that shouldly is going to be your performance bottleneck. We could put the noinline attribute on all methods in `ShouldBeStringTestExtensions`, that would fix the case I'm testing, but maybe not all cases.

**Comment by slang25** [comment]
Side story, I once fixed a production bug by flipping that bit in production 😆 bad times

**Comment by jnm2** [comment]
Jan Kotas says:

> The runtime treats `MethodImplOptions.NoInlining` like PreserveStackFrame: https://github.com/dotnet/coreclr/blob/master/src/vm/jitinterface.cpp#L8308.

The linked code:

> ```cpp
> // Do not tailcall from methods that are marked as noinline (people often use no-inline
> // to mean "I want to always see this method in stacktrace")
> ```

That comment has been there since the CoreCLR repo was first populated from .NET Framework, so I imagine NoInlining disables both inlining and tail calls on both .NET Framework and .NET Core. No idea about Mono.

**Comment by JosephWoodward** [comment]
Oh nice find @jnm2! Thanks for the link.

**Comment by slang25** [comment]
What also tripped me up while debugging this was I thought that this code was getting hit:
https://github.com/shouldly/shouldly/blob/master/src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs
However, there is a completely different implementation that is actually used here:
https://github.com/shouldly/shouldly/blob/ea5739a3859627d4112723c1ddab557d1b842e1b/src/Shouldly/Internals/SourceCodeTextGetter.cs#L60
I'm still looking at fixing the logic with simple simple yet robust, not there yet though.

**Comment by JosephWoodward** [comment]
@thomasthiry @jnm2 @slang25  I've just [released version 3.0.2](https://www.nuget.org/packages/Shouldly/3.0.2), thanks for your help with this 👍

**Comment by thomasthiry** [comment]
Yoohoo it works! :-)

Before:
```
Shoudly.ShouldAssertException : 4
    should be
1
    but was not
```

After:
```
Shouldly.ShouldAssertException : employeeModel.NbExtraHolidaysDays
    should be
1
    but was
4
```

Thanks a lot for all the good work! Now I can use it for my workshops and show everybody how easy it is. :-)

**Comment by JosephWoodward** [comment]
Yay! Glad to see it's working now, thanks for reporting it :)
